Failed to build `cuda_ukernel_ukernel.cu.bc` on Windows

### What happened?

Tried to build after syncing on my Windows machine, got these error logs:

```
[build] -- Using found CUDA toolkit: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.6
[build] -- Using CUDA libdevice: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.6/bin/../nvvm/libdevice/libdevice.10.bc
[build] -- Using CUDA INCLUDE_DIRS from found SDK: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.6/include

...

[build] [140/745  10% :: 1.699] Compiling ukernel.cu to cuda_ukernel_ukernel.cu.bc
[build] FAILED: samples/custom_dispatch/cuda/kernels/cuda_ukernel_ukernel.cu.bc D:/dev/projects/iree-build/samples/custom_dispatch/cuda/kernels/cuda_ukernel_ukernel.cu.bc 
[build] cmd.exe /C "cd /D D:\dev\projects\iree-build\samples\custom_dispatch\cuda\kernels && D:\dev\projects\iree-build\llvm-project\bin\clang.exe -x cuda --cuda-gpu-arch=sm_60 -Wno-unknown-cuda-version -nocudalib --cuda-device-only -O3 -c -emit-llvm D:/dev/projects/iree/samples/custom_dispatch/cuda/kernels/ukernel.cu -o cuda_ukernel_ukernel.cu.bc"
[build] clang: error: cannot find CUDA installation; provide its path via '--cuda-path', or pass '-nocudainc' to build without CUDA includes
```

### Steps to reproduce your issue

1. Configure CMake on Windows / MSVC with `-DIREE_HAL_DRIVER_CUDA=ON -DIREE_TARGET_BACKEND_CUDA=ON`
2. Try building `all`
3. See error


### What component(s) does this issue relate to?

Compiler

### Version information

Tip of tree (https://github.com/openxla/iree/commit/2f2d28c09e08a78f05b00e82707fa66f2fdf99a4)

### Additional context

```
λ cmake --version
cmake version 3.24.0

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html

> `CUDAToolkit_LIBRARY_ROOT`
New in version 3.18.

```
[build] -- CUDAToolkit_FOUND
[build] -- CUDAToolkit_BIN_DIR: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.6/bin
[build] -- CUDAToolkit_LIBRARY_ROOT: 
[build] -- CUDAToolkit_LIBRARY_DIR: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.6/lib/x64
```

Relevant source:
* https://github.com/openxla/iree/blob/2f2d28c09e08a78f05b00e82707fa66f2fdf99a4/build_tools/cmake/iree_bitcode_library.cmake#L142-L146
* https://github.com/openxla/iree/blob/2f2d28c09e08a78f05b00e82707fa66f2fdf99a4/samples/custom_dispatch/cuda/kernels/CMakeLists.txt#L90-L94

Tried

* Updating my CUDA Toolkit to v12.2
* Adding `"--cuda-path=${CUDAToolkit_ROOT}"` to `build_tools\cmake\iree_bitcode_library.cmake` (sort of, the original variable in `build_tools\third_party\cuda\CMakeLists.txt` only sets it in local scope)

Got a new error:

```
[build] FAILED: samples/custom_dispatch/cuda/kernels/cuda_ukernel_ukernel.cu.bc D:/dev/projects/iree-build/samples/custom_dispatch/cuda/kernels/cuda_ukernel_ukernel.cu.bc 
[build] cmd.exe /C "cd /D D:\dev\projects\iree-build\samples\custom_dispatch\cuda\kernels && D:\dev\projects\iree-build\llvm-project\bin\clang.exe -x cuda --cuda-gpu-arch=sm_60 "--cuda-path=C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.2" -Wno-unknown-cuda-version -nocudalib --cuda-device-only -O3 -c -emit-llvm D:/dev/projects/iree/samples/custom_dispatch/cuda/kernels/ukernel.cu -o cuda_ukernel_ukernel.cu.bc"
[build] In file included from <built-in>:1:
[build] In file included from D:\dev\projects\iree-build\llvm-project\lib\clang\17\include\__clang_cuda_runtime_wrapper.h:41:
[build] In file included from D:\dev\projects\iree-build\llvm-project\lib\clang\17\include\cuda_wrappers\cmath:27:
[build] In file included from C:\Program Files (x86)\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.31.31103\include\cmath:9:
[build] In file included from C:\Program Files (x86)\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.31.31103\include\yvals.h:9:
[build] C:\Program Files (x86)\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.31.31103\include\yvals_core.h:581:2: error: STL1002: Unexpected compiler version, expected CUDA 10.1 Update 2 or newer.
[build]   581 | #error STL1002: Unexpected compiler version, expected CUDA 10.1 Update 2 or newer.
[build]       |  ^
[build] 1 error generated when compiling for sm_60.
```

--- a/build_tools/cmake/iree_bitcode_library.cmake
+++ b/build_tools/cmake/iree_bitcode_library.cmake
@@ -132,18 +132,23 @@ function(iree_cuda_bitcode_library)
   endif()
 
   set(_CUDA_ARCH "${_RULE_CUDA_ARCH}")
-  
+
   set(_COPTS
     "-x" "cuda"
-    
+
     # Target architecture.
     "--cuda-gpu-arch=${_CUDA_ARCH}"
+
+    "--cuda-path=${CUDAToolkit_ROOT}"
 
     # Suppress warnings about missing path to cuda lib,
     # and benign warning about CUDA version.
     "-Wno-unknown-cuda-version"
     "-nocudalib"
     "--cuda-device-only"
+
+    # https://github.com/llvm/llvm-project/issues/54609
+    "-D_ALLOW_COMPILER_AND_STL_VERSION_MISMATCH"
 
     # Optimized and unstamped.
     "-O3"

--- a/samples/custom_dispatch/cuda/kernels/CMakeLists.txt
+++ b/samples/custom_dispatch/cuda/kernels/CMakeLists.txt
@@ -134,7 +134,7 @@ iree_cuda_bitcode_library(
   CUDA_ARCH
     sm_60
   SRCS
-    "ukernel.cu" 
+    "ukernel.cu"
 )
 
 iree_check_single_backend_test_suite(

